src/modules/musig/main_impl.h

```diff
@@ -45,64 +45,58 @@ static void secp256k1_musig_sha256_init_tagged(secp256k1_sha256 *sha) {
     sha->bytes = 64;
 }
 
-/* Compute r = SHA256(ell, idx). The four bytes of idx are serialized least significant byte first. */
-static void secp256k1_musig_coefficient(secp256k1_scalar *r, const unsigned char *ell, uint32_t idx) {
+/* Compute MuSig coefficient which is constant 1 for the second pubkey and
+ * SHA256(ell, x) otherwise. second_pk_x can be 0 in case there is no
+ * second_pk. Assumes both field elements x and second_pk_x are normalized. */
+static void secp256k1_musig_coefficient_internal(secp256k1_scalar *r, const unsigned char *ell, const secp256k1_fe *x, const secp256k1_fe *second_pk_x) {
     secp256k1_sha256 sha;
     unsigned char buf[32];
-    size_t i;
 
-    secp256k1_musig_sha256_init_tagged(&sha);
-    secp256k1_sha256_write(&sha, ell, 32);
-    /* We're hashing the index of the signer instead of its public key as specified
-     * in the MuSig paper. This reduces the total amount of data that needs to be
-     * hashed.
-     * Additionally, it prevents creating identical musig_coefficients for identical
-     * public keys. A participant Bob could choose his public key to be the same as
-     * Alice's, then replay Alice's messages (nonce and partial signature) to create
-     * a valid partial signature. This is not a problem for MuSig per se, but could
-     * result in subtle issues with protocols building on threshold signatures.
-     * With the assumption that public keys are unique, hashing the index is
-     * equivalent to hashing the public key. Because the public key can be
-     * identified by the index given the ordered list of public keys (included in
-     * ell), the index is just a different encoding of the public key.*/
-    for (i = 0; i < sizeof(uint32_t); i++) {
-        unsigned char c = idx;
-        secp256k1_sha256_write(&sha, &c, 1);
-        idx >>= 8;
-    }
-    secp256k1_sha256_finalize(&sha, buf);
-    secp256k1_scalar_set_b32(r, buf, NULL);
+    if (secp256k1_fe_cmp_var(x, second_pk_x) == 0) {
+        secp256k1_scalar_set_int(r, 1);
+    } else {
+        secp256k1_musig_sha256_init_tagged(&sha);
+        secp256k1_sha256_write(&sha, ell, 32);
+        secp256k1_fe_get_b32(buf, x);
+        secp256k1_sha256_write(&sha, buf, 32);
+        secp256k1_sha256_finalize(&sha, buf);
+        secp256k1_scalar_set_b32(r, buf, NULL);
+    }
+}
+
+static void secp256k1_musig_coefficient(secp256k1_scalar *r, const secp256k1_musig_pre_session *pre_session, secp256k1_fe *x) {
+    secp256k1_fe second_pk_x;
+    secp256k1_fe_set_b32(&second_pk_x, pre_session->second_pk);
+    secp256k1_musig_coefficient_internal(r, pre_session->pk_hash, x, &second_pk_x);
```

Comment:
`secp256k1_musig_coefficient_internal` assume that `second_pk_x` is normalized however, I don't see any normalization before affectation of `pre_session->second_pk`.

Reply:
similarly https://github.com/ElementsProject/secp256k1-zkp/blob/f7a27e6cca0116f60c5d1a6d686e262e8fe1fbf7/src/modules/musig/main_impl.h#L88 does not seems to check that `x` is normalized.

Reply:
The output of `fe_set_b32` and the coordinates of the output of `pubkey_load` are always normalized